controller/state.go

```diff
@@ -651,6 +662,67 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *ap
 	return &compRes
 }
 
+type normalizeResults struct {
+	lives      []*unstructured.Unstructured
+	targets    []*unstructured.Unstructured
+	conditions []v1alpha1.ApplicationCondition
+}
+
+type preDiffNormalizeParams struct {
+	targets          []*unstructured.Unstructured
+	lives            []*unstructured.Unstructured
+	ignoreDiffConfig *argo.IgnoreDiffConfig
+	appLabel         string
+	trackingMethod   string
+}
+
+// preDiffNormalize applies the normalization of live and target resources before invoking
+// the diff. None of the attributes in the preDiffNormalizeParams will be modified. The
+// normalizeResults will return a list of ApplicationConditions in case something goes
+// wrong during the normalization.
+func (m *appStateManager) preDiffNormalize(p *preDiffNormalizeParams) *normalizeResults {
+	results := &normalizeResults{}
+	for i := range p.targets {
+		target := safeDeepCopy(p.targets[i])
+		live := safeDeepCopy(p.lives[i])
+		_ = m.resourceTracking.Normalize(target, live, p.appLabel, p.trackingMethod)
```

Comment:
I am not sure if the input of `resourceTracking.Normalize` should be the original value or the copy.

Reply:
As we discussed, this change also fix another bug where the code was mutating the target resource and it was reflected in the sync when it shouldn't. The new code will never mutate the objects that are used during the sync